mage_ai/tests/ai/test_ai_functions.py

```diff
@@ -0,0 +1,133 @@
+
+EXPECTED_RESPONSE_FOR_LLM_SPLIT = """
+Answer:
+BLOCK 1: function: load data from MySQL. upstream:
+BLOCK 2: function: filter out records column age < 18. upstream: 1
+BLOCK 3: function: export it to postgres. upstream: 2"""
+EXPECTED_LOAD_BLOCK = '{"code": "load mysql"}'
+EXPECTED_FILTER_BLOCK = '{"code": "filter record"}'
+EXPECTED_EXPORT_BLOCK = '{"code": "export postgres"}'
+MOCK_TRANSFORMER_CLASSIFICATION = {
+    "role": "assistant",
+    "function_call": {
+        "name": "classify_description",
+        "arguments": "{\n  \"BlockType\": \"BlockType__transformer\",\n  "
+        "\"BlockLanguage\": \"BlockLanguage__python\",\n  "
+        "\"ActionType\": \"ActionType__filter\"\n}"
+    }
+}
+MOCK_LOADER_FUNCTION_CLASSIFICATION = {
+    "role": "assistant",
+    "function_call": {
+        "name": "classify_description",
+        "arguments": "{\n  \"BlockType\": \"BlockType__data_loader\",\n  "
+        "\"BlockLanguage\": \"BlockLanguage__python\",\n  "
+        "\"DataSource\": \"DataSource__mysql\"\n}"
+    }
+}
+MOCK_EXPORTERS_FUNCTION_CLASSIFICATION = {
+    "role": "assistant",
+    "function_call": {
+        "name": "classify_description",
+        "arguments": "{\n  \"BlockType\": \"BlockType__data_exporter\",\n  "
+        "\"BlockLanguage\": \"BlockLanguage__python\",\n  "
+        "\"DataSource\": \"DataSource__postgres\"\n}"
+    }
+}
+
+
+class AIFunctionTest(TestCase):
+    def setUp(self):
+        self.wizard = LLMPipelineWizard()
+        self.pipeline = Pipeline.create("test_pipeline", repo_path="test")
+        self.block = Block.create(name="test_block", block_type="data_loader", repo_path="test")
+        self.pipeline.add_block(self.block)
+        self.loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(self.loop)
+
+    def tearDown(self):
+        self.pipeline.delete()
+        self.loop.close()
+
+    def __set_expected_future_result(self, expected_value):
+        result_holder = asyncio.Future()
+        result_holder.set_result(expected_value)
+        return result_holder
+
+    def test_async_generate_pipeline_from_description(self):
+        mock_customized_code = self.__set_expected_future_result({"action_code": "filter"})
+        self.wizard._LLMPipelineWizard__async_llm_call = MagicMock(
+            side_effect=[self.__set_expected_future_result(EXPECTED_RESPONSE_FOR_LLM_SPLIT),
+                         mock_customized_code,
+                         mock_customized_code,
+                         mock_customized_code])
+        self.wizard._LLMPipelineWizard__async_identify_function_parameters = MagicMock(
+            side_effect=[
+                self.__set_expected_future_result(MOCK_LOADER_FUNCTION_CLASSIFICATION),
+                self.__set_expected_future_result(MOCK_TRANSFORMER_CLASSIFICATION),
+                self.__set_expected_future_result(MOCK_EXPORTERS_FUNCTION_CLASSIFICATION)])
+
+        pipeine_description = "load data from mysql, filter out "
```

Comment:
for multiline string, you'll need to put `\` at the end of the line
a = "aaaaa" \ 
  "bbbbb"

Reply:
Updated